pubsub/gossipsub/gossipsub-v1.2.md

```diff
@@ -4,16 +4,103 @@
+ peer penalizations
+
+| Parameter                | Description                                                      | Reasonable Default |
+|--------------------------|------------------------------------------------------------------|--------------|
+| `max_idontwant_messages` | The maximum number of `IDONTWANT` messages per heartbeat per peer | ???  |
+
+
+## IDONTWANT Message
+
+### Basic scenario
+
+When the peer receives the first message instance it immediately broadcasts 
+(not queue for later piggybacking) `IDONTWANT` with the `messageId` to all its mesh peers. 
+This could be performed prior to the message validation to further increase the effectiveness of the approach.    
+
+On the other side a node maintains per-peer `dont_send_message_ids` set. Upon receiving `IDONTWANT` from 
+a peer the `messageId` is added to the `dont_send_message_ids` set. 
+When later relaying the `messageId` message to the mesh the peers found in `dont_send_message_ids` MUST be skipped. 
+
+Old entries from `dont_send_message_ids` SHOULD be pruned during heartbeat processing. 
+The prune strategy is outside of the spec scope and can be decided by implementations.
+
+`IDONTWANT` message is supposed to be _optional_ for both receiver and sender. I.e. the sender MAY NOT utilize 
+this message. The receiver in turn MAY ignore `IDONTWANT`: sending a message after the corresponding `IDONTWANT` 
+should not be penalized.    
+
+The `IDONTWANT` may have negative effect on small messages as it may increase the overall traffic and CPU load.
+Thus it is better to utilize `IDONTWANT` for messages of a larger size.
+The exact policy of `IDONTWANT` appliance is outside of the spec scope. Every implementation MAY choose whatever 
+is more appropriate for it. Possible options are either choose a message size threshold and broadcast `IDONTWANT`
+on per message basis when the size is exceeded or just use `IDONTWANT` for all messages on selected topics.
+
+To prevent DoS the number of `IDONTWANT` control messages is limited to `max_idontwant_messages` per heartbeat  
+
+### Cancelling `IWANT`
+
+If a node requested a message via `IWANT` and then occasionally receives the message from other peer it MAY 
+try to cancel its `IWANT` requests with the corresponding `IDONTWANT` message. It may work in cases when a
+peer delays/queues `IWANT` requests and the `IWANT` request SHOULD be removed from the queue if not processed yet
+
+## Protobuf Extension
+
+The protobuf messages are identical to those specified in the [gossipsub v1.0.0
+specification](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.0.md)
+with the following  control message modifications:
+
+```protobuf
+message RPC {
+ // ... see definition in the gossipsub specification
+}
+
+message ControlMessage {
+    // messages from v1.0
+    repeated ControlIDontWant iDontWant = 5;
+}
+
+message ControlIDontWant {
+    required bytes messageID = 1;
```

Comment:
Why not making it plural just like `IHAVE` and `IWANT`? I know that it's unnecessary because you can just put many `ControlIDontWant`s into the message, but isn't it better to make it consistent with `IHAVE` and `IWANT`?
```suggestion
    repeated bytes messageIDs = 1;
```

Reply:
Agreed, thats better.

Reply:
Yep, sounds reasonable 👍  Thanks for comment!